Automattic/woocommerce-payments
# Decouple the 'force save payment method' flag from payment metadata

Fixes #861

#### Changes proposed in this Pull Request

* Introduce the `WCPay\Constants\Payment_Type` class to isolate payment type constants into a separate file.
* Use the newly introduced constants within `process_payment` (both standard and subs) in order to assign the right payment type in the right situation.
* Fix [`WC_REST_Payments_Webhook_Controller_Test::test_valid_failed_refund_update_webhook`](https://github.com/Automattic/woocommerce-payments/pull/979/commits/3c8ea2229f3718f4afd242ac7b810fd782924428) in order to avoid failing tests with the latest WooCommerce Version.
* Add unit tests for the payment type meta, because it's only sent to Stripe's API and failure would be easy to miss.

#### Testing instructions

1. Purchase a normal product.
2. Check the latest API log and make sure that the `payment_type` is set to `single` in the latest request's metadata.
3. Purchase a subscription product.
4. Check the latest API log and make sure that the `payment_type` is set to `recurring` in the latest request's metadata.
5. In the admin, renew a subscription.
6. Check the latest API log and make sure that the `payment_type` is set to `single` in the latest request's metadata.

#### Note

This is a minor change and I'm not adding a changelog entry for it, as it doesn't affect merchants.

-------------------
